qpc/report/test_report_merge.py

```diff
@@ -73,295 +120,324 @@ def setup_class(cls):
+        details2,
+        bad_details1,
+        bad_details2,
+        bad_details3,
+        bad_details4,
+        bad_details5,
+        gooddetails,
+        not_json_file,
+        capsys,
+        requests_mock,
+        caplog,
+    ):
+        """Testing report merge command with json directory."""
         args = Namespace(
             scan_job_ids=None,
             json_files=None,
             report_ids=None,
-            json_dir=bad_json_directory,
+            json_dir=[str(Path(details1).parent)],
         )
-        with pytest.raises(SystemExit):
-            with redirect_stdout(report_out):
-                self.command.main(args)
-
-    def test_detail_merge_json_directory_error_path_passed(self):
-        """Testing report merge command with json_dir parameter (file)."""
-        report_out = StringIO()
-
-        args = Namespace(
-            scan_job_ids=None,
-            json_files=None,
-            report_ids=None,
-            json_dir=TMP_BADDETAILS1[0],
+        requests_mock.post(
+            get_server_location() + ASYNC_MERGE_URI, status_code=201, json={"id": 1}
         )
-        with pytest.raises(SystemExit):
-            with redirect_stdout(report_out):
-                self.command.main(args)
-
-    def test_detail_merge_json_invalid_report_type_passed(self):
-        """Testing report merge command bad report_type."""
-        report_out = StringIO()
-
+        caplog.set_level(logging.ERROR)
+        self.command.main(args)
+        expected_msg = messages.REPORT_SUCCESSFULLY_MERGED % {
+            "id": "1",
+            "prog_name": QPC_VAR_PROGRAM_NAME,
+        }
+        assert expected_msg in capsys.readouterr().out
+        # ensure only good reports were pushed to server
+        assert requests_mock.last_request.json() == {
+            "report_type": "details",
+            "sources": [mock.ANY, mock.ANY],
+        }
+        assert {
+            "facts": ["A"],
+            "report_type": "details",
+            "report_version": "1.2.3+outer_version",
+            "server_id": "42",
+        } in requests_mock.last_request.json()["sources"]
+        assert {
+            "facts": ["B"],
+            "report_type": "details",
+            "report_version": "0.0.44.legacy",
+            "server_id": "8",
+        } in requests_mock.last_request.json()["sources"]
+        # ensure bad reports were detected
+        assert (
+            messages.REPORT_JSON_MISSING_ATTR % {"file": details2, "key": "facts"}
+            in caplog.text
+        )
+        assert (
+            messages.REPORT_JSON_MISSING_ATTR % {"file": bad_details1, "key": "sources"}
+            in caplog.text
+        )
+        assert (
+            messages.REPORT_JSON_MISSING_ATTR % {"file": bad_details2, "key": "facts"}
+            in caplog.text
+        )
+        assert (
+            messages.REPORT_JSON_MISSING_ATTR % {"file": bad_details3, "key": "facts"}
+            in caplog.text
+        )
```

Comment:
the error message highlights that bad_details2 and 3 probably need to be redesigned

Reply:
Add this as an inline comment for future readers? I appreciate the other inline comments you've added here around other weird behaviors.